java/src/com/ibm/streamsx/topology/generator/spl/Optimizer.java

```diff
@@ -0,0 +1,136 @@
+
+    static {
+        for (String kind : new String[] { "Source", "Filter", "Map", "FlatMap", "ForEach"}) {
+            PY_FUNC_OPS.add(PY_OP_NS + "::" + kind);
+            PY_FUNC_OPS.add(PY_OP_NS + "2::" + kind);
+        }
+    }
+
+    /**
+     * Setup Python operators to allow pass by reference.
+     * 
+     * Finds Python functional operators and sets the outputConnections
+     * parameter representing the number of connections.
+     * 
+     * If pass by reference cannot be used outputConnections will not be set.
+     * 
+     * Does not modify the structure of the graph.
+     * Assumes the graph's structure will not be subsequently modified.
+     */
+    private final void pyPassByRef() {
+        Set<JsonObject> pyops = findOperatorsByKinds(graph, PY_FUNC_OPS);
+
+        if (pyops.isEmpty())
+            return;
+
+        for (JsonObject pyop : pyops) {
+            JsonArray outputs = array(pyop, "outputs");
+            if (outputs == null || outputs.size() == 0)
+                continue;
+
+            // Currently only supporting a single output port
+            // though mostly coded to support N.
+            assert outputs.size() == 1;
+
+            int[] connCounts = new int[outputs.size()];
+
+            for (int port = 0; port < connCounts.length; port++) {
+                connCounts[port] = -1;
+
+                JsonObject output = outputs.get(port).getAsJsonObject();
+
+                // Can't use the schema objects as we need to not depend on IBM Streams
+                // classes.
+                if (!"tuple<blob __spl_po>".equals(jstring(output, "type")))
+                    continue;
+
+                JsonArray conns = array(output, "connections");
+                if (conns == null || conns.size() == 0) {
+                    connCounts[port] = 0;
+                    continue;
+                }
+                
+                boolean canPassByRef = true;
+                // TOOD - downstream for a specific port
+                Set<JsonObject> connected = getDownstream(pyop, graph);
+                for (JsonObject connectedOp : connected) {
+                    if (!PY_FUNC_OPS.contains(kind(connectedOp))) {
+                        canPassByRef = false;
+                        break;
+                    }
+                        
+                        // TEMP
+                        // Currently only Map and ForEach completly handle
+                        // by reference.
+                        if (!kind(connectedOp).endsWith("::Map") && !kind(connectedOp).endsWith("::ForEach")) {
+                        canPassByRef = false;
+                        break;
+                        }
+                }
+
+                if (canPassByRef)
+                    connCounts[port] = conns.size();
+            }
+
+            boolean paramNeeded = false;
+            for (int i = 0; i < connCounts.length; i++) {
+                if (connCounts[0] != -1) {
```

Comment:
Since we only support one output port right now, I don't think this breaks anything yet.